File: PROTOTIPO_C/TA 033 EA2 E9.c
//2101295ghgh2101295hghghghgelvaelva

#include <stdio.h>
#include <string.h>

void mostrar_datos();
int leer_cadena();
int validar_cadena(char *cadena);
int validar_i(char *cadena, char *i, int *index);
int validar_w(char *cadena, char *w,int *index, int *n);
int validar_wi(char *cadena, char *wi,int *index, int *n);
int validar_j(char *cadena, char *j, int *index);
int leer_opcion();

int main (){
	int opc = 1;
	while(opc == 1){
		system("cls");
		mostrar_datos();
		if(leer_cadena())
			printf("\nLA CADENA ES VALIDA PARA ESTA GRAMATICA\n");
		else
			printf("\nLA NO CADENA ES VALIDA PARA ESTA GRAMATICA\n");
		printf("\nDesea ingresar otra cadena?\n1-SI\n2-NO\n-> ");
		opc = leer_opcion();
	}
	return 0;
}

int leer_cadena(){
	char cadena[100];
	printf("\nINGRESE UNA CADENA\n");
	fgets(cadena, sizeof(cadena), stdin);
	cadena[strcspn(cadena, "\n")] = '\0';
	if (cadena[0] == '\0'){
		return 0;	
	}
	if(validar_cadena(cadena)){
		return 1;
	}
	else return 0;
}

int validar_cadena(char *cadena){
	char w[] = "gh";
	char wi[] = "hg";
	char j[] =  "elva";
	char i[] = "2101295";
	int n = 0;
	int index = 0;
	if(!validar_i(cadena, i, &index)){
		printf("\nCadena no valida, la cadena i no se encuentra al inicio\n");
		return 0;
	}
	if(!validar_w(cadena, w, &index, &n)){
		printf("\nCadena no valida, la cadena w debe aparecer al menos una vez\n");
		return 0;
	}
	if(!validar_i(cadena, i, &index)){
		printf("\nCadena no valida, la cadena i debe de aparecer despues de las n repeticiones de w\n");
		return 0;
	}
	if(!validar_wi(cadena, wi, &index, &n)){
		printf("\nCadena no valida, la cadena w debe aparecer el doble de veces de w\n");
		return 0;
	}
	if(!validar_j(cadena, j, &index)){
		printf("\nCadena no valida, la cadena j debe aparecer 2 veces\n");
		return 0;
	}
	if (cadena[index] != '\0') return 0;
	return 1;
}

int validar_i(char *cadena, char *i, int *index){
	int j;
	for (j = 0; j < strlen(i); j++){
		if(cadena[(*index)] == i[j]){
			(*index)++;
		}
		else
			return 0;
	}
	return 1;
}

int validar_w(char *cadena, char *w,int *index, int *n){
	while (cadena[(*index)] == w[0] && cadena[(*index) + 1] == w[1]){
		(*n)++;
		(*index) += 2;
	}
	if((*n) == 0)
		return 0;
	return 1;
}

int validar_wi(char *cadena, char *wi,int *index, int *n){
	int ni = 0;
	while (cadena[*index] == wi[0] && cadena[(*index) + 1] == wi[1]){
		ni++;
		(*index) += 2;
	}
	if(ni != 2 * (*n))
		return 0;
	return 1;
}

int validar_j(char *cadena, char *j, int *index){
    int len = strlen(j);
	int i;
    for ( i = 0; i < len; i++){
        if (cadena[*index + i] != j[i])
            return 0;
    }
    for ( i = 0; i < len; i++){
        if (cadena[*index + len + i] != j[i])
            return 0;
    }
    *index += 2 * len;
    return 1;
}

void mostrar_datos(){
	printf("\nVALIDAR UNA CADENA A PARTIR DE UNA GRAMATICA (GIC)\n");
	printf("\ngramatica: G = {I, W^n, I, WI^2n, J^2 | n >= 1}\n");
	printf("I = 2101295\nW = gh\nWI = hg\nJ = elva\n");
}

int leer_opcion(){
	int valor;
	int valido = 0;
	do{
		if (scanf("%d", &valor) == 1 && (valor == 2 || valor == 1))
			valido = 1;
		else{
			printf("\nIngrese una opcion valida\n-> ");
			valido = 0;
		}
		while (getchar() != '\n');
	} while (!valido);

	return valor;
}

File: PROTOTIPO_PYTHON/TA 033 EA2 E9.py
def validar_i(cadena, i, index):
    for j in range(len(i)):
        if index >= len(cadena) or cadena[index] != i[j]:
            return None
        index += 1
    return index

def validar_w(cadena, w, index):
    n = 0
    while cadena[index:index+len(w)] == w:
        n += 1
        index += len(w)
    if n == 0:
        return None, None
    return index, n

def validar_wi(cadena, wi, index, n):
    ni = 0
    while cadena[index:index+len(wi)] == wi:
        ni += 1
        index += len(wi)
    if ni != 2 * n:
        return None
    return index


def validar_j(cadena, j, index):
    len_j = len(j)
    for i in range(len_j):
        if index + i >= len(cadena) or cadena[index + i] != j[i]:
            return None
    for i in range(len_j):
        if index + len_j + i >= len(cadena) or cadena[index + len_j + i] != j[i]:
            return None
    return index + 2 * len_j


def validar_cadena(cadena):
    w = "gh"
    wi = "hg"
    j = "elva"
    i = "2101295"
    index = 0

    index = validar_i(cadena, i, index)
    if index is None:
        print("Error: I no está al inicio")
        return False

    res = validar_w(cadena, w, index)
    if res[0] is None:
        print("Error: w debe aparecer al menos una vez")
        return False
    index, n = res

    index = validar_i(cadena, i, index)
    if index is None:
        print("Error: falta la segunda I")
        return False

    index = validar_wi(cadena, wi, index, n)
    if index is None:
        print("Error: w inversa no cumple 2n")
        return False

    index = validar_j(cadena, j, index)
    if index is None:
        print("Error: j debe aparecer 2 veces")
        return False
    if index != len(cadena):
        print("Error: hay caracteres extra")
        return False

    return True

def mostrar_datos():
    print("\nVALIDAR UNA CADENA A PARTIR DE UNA GRAMATICA (GIC)")
    print("\ngramatica: G = {I, W^n, I, WI^2n, J^2 | n >= 1}")
    print("I = 2101295")
    print("W = gh")
    print("WI = hg")
    print("J = elva")

def leer_opcion():
    while True:
        try:
            valor = int(input("-> "))
            if valor in (1, 2):
                return valor
            else:
                print("Ingrese una opción válida")
        except:
            print("Ingrese un número válido")

opc = 1
while opc == 1:
    print("\n" * 3)
    mostrar_datos()

    cadena = input("\nINGRESE UNA CADENA:\n")

    if cadena == "":
        print("\nCadena inválida (vacía)")
    else:
        if validar_cadena(cadena):
            print("\nLA CADENA ES VALIDA PARA ESTA GRAMATICA")
        else:
            print("\nLA CADENA NO ES VALIDA PARA ESTA GRAMATICA")

    print("\nDesea ingresar otra cadena?")
    print("1 - SI")
    print("2 - NO")

    opc = leer_opcion()